substratools/algo.py

```diff
@@ -104,312 +67,96 @@ def _parser_add_default_arguments(parser):
+    def task_launcher(self, method_name: str, rank: int = 0, fake_data: bool = False, n_fake_samples: int = None):
+        """Launch Tasks"""
 
-        # load models
-        models = self._load_models_paths()
+        # load inputs
+        inputs = deepcopy(self._workspace.task_inputs)
 
-        # train new model
-        logger.info("launching training task")
-        method = self._interface.train if not fake_data else self._interface._train_fake_data
+        inputs.update({"rank": rank})
 
-        # TODO get data and models labels from --inputs
-        method(
-            inputs={
-                InputIdentifiers.X: X,
-                InputIdentifiers.y: y,
-                InputIdentifiers.models: models,
-                InputIdentifiers.rank: rank,
-            },
-            outputs={OutputIdentifiers.model: self._workspace.output_model_path},
-        )
+        # load data from opener
+        if self._opener_wrapper:
+            X = self._opener_wrapper.get_X(fake_data, n_fake_samples)
+            y = self._opener_wrapper.get_y(fake_data, n_fake_samples)
 
-        self._assert_output_exists(self._workspace.output_model_path, OutputIdentifiers.model)
+            if fake_data:
+                logger.info("Using fake data with %i fake samples." % int(n_fake_samples))
 
-    @utils.Timer(logger)
-    def predict(self, fake_data=False, n_fake_samples=None, **kwargs):
-        """Predict method wrapper."""
-        # load data from opener
-        X = self._opener_wrapper.get_X(fake_data, n_fake_samples)
+            inputs.update({"X": X})
+            inputs.update({"y": y})
 
-        # load models
-        model_paths = self._workspace.input_model_paths
-        if not model_paths or len(model_paths) != 1:
-            raise exceptions.InvalidInputOutputsError("predict expects exactly one input model")
-        model_path = model_paths[0]
+        # load outputs
+        outputs = deepcopy(self._workspace.task_outputs)
 
-        # get predictions
-        logger.info("launching predict task")
-        method = self._interface.predict if not fake_data else self._interface._predict_fake_data
-        inputs = {InputIdentifiers.X: X, InputIdentifiers.model: model_path}
-        outputs = {OutputIdentifiers.predictions: self._workspace.output_predictions_path}
+        # Retrieve method from user
+        method = getattr(self._interface, method_name)
 
-        method(inputs, outputs)
+        logger.info("Launching task: executing `%s` function." % method_name)
+        method(
+            inputs=inputs,
+            outputs=outputs,
+        )
 
-        self._assert_output_exists(self._workspace.output_predictions_path, OutputIdentifiers.predictions)
+        self._assert_outputs_exists(
+            self._workspace.task_outputs,
+        )
 
 
-def _generate_algo_cli(interface):
+def _generate_generic_algo_cli(interface):
     """Helper to generate a command line interface client."""
 
     def _algo_from_args(args):
```

Comment:
The name of the function implies that it builds an algo structure from its args. But it also has 2 side effects:

- it configures logging globally
- it loads a module (also a global operation with side effects)

I'd recommend moving statements with side effects outside of `_algo_from_args`

Reply:
This has been added to the backlog for a future PR ("Refactor substra-tools")